Use more smart pointers in CSSStyleSheet

zakariaridouh opened the pull request:
#### f5dc95002e132c49f9f91d8ff269dde4ae90f2fc
<pre>
Use more smart pointers in CSSStyleSheet
<a href="https://bugs.webkit.org/show_bug.cgi?id=278263">https://bugs.webkit.org/show_bug.cgi?id=278263</a>
&lt;<a href="https://rdar.apple.com/134095002">rdar://134095002</a>&gt;

Reviewed by Matthieu Dubet.

Use some more smart pointers in CSSStyleSheet. This is the first of a
series of PRs in which I will adopt more smart pointer usage in respect
to the WebKit Smart Pointer style guide, starting with the `WebCore/css`
folder here.

* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::~CSSStyleSheet):
(WebCore::CSSStyleSheet::RuleMutationScope::RuleMutationScope):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/StyleSheet.h:
(WebCore::StyleSheet::parentStyleSheet const):
(WebCore::StyleSheet::media const):
(WebCore::StyleSheet::ownerRule const):

Canonical link: <a href="https://commits.webkit.org/293318@main">https://commits.webkit.org/293318@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e78b604b5d9abedbde94b1240474940b32c2852e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98537 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18170 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8403 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103658 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49078 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100581 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18463 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26621 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/75013 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32169 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101541 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/14007 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88993 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55371 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/13786 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6958 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48507 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83742 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7036 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106031 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25627 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/18664 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83984 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26004 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85194 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83469 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21088 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28113 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5784 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19308 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25585 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30766 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25403 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28723 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26978 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rreno commented on Source/WebCore/css/CSSStyleSheet.cpp:
```diff
@@ -71,9 +71,9 @@ class StyleSheetCSSRuleList final : public CSSRuleList {
+    CSSStyleSheet* styleSheet() const final { return m_styleSheet.get(); }
 
-    CSSStyleSheet* m_styleSheet;
+    RefPtr<CSSStyleSheet> m_styleSheet;
```
This is likely a reference cycle. This can get stored as a member variable on the CSSStyleSheet itself. I'd suggest using a `WeakPtr` instead.

rreno replied:
This being the `StyleSheetCSSRuleList` object.

zakariaridouh replied:
Thank you for explaining. Fixed.

rreno reviewed (commented):
LGTM but you should probably get someone who knows Style more than I do to give you the final r+

mdubet commented on Source/WebCore/css/CSSStyleSheet.cpp:
```diff
@@ -71,9 +71,9 @@ class StyleSheetCSSRuleList final : public CSSRuleList {
+    CSSStyleSheet* styleSheet() const final { return m_styleSheet.get(); }
 
-    CSSStyleSheet* m_styleSheet;
+    SingleThreadWeakPtr<CSSStyleSheet> m_styleSheet;
```
Why do we use the `SingleThreadWeakPtr` here instead of the general `WeakPtr` ?

zakariaridouh replied:
`CSSStyleSheet` only inherits from `CanMakeSingleThreadWeakPtr<CSSStyleSheet>`, so it needs to be a `SingleThreadWeakPtr`

https://github.com/WebKit/WebKit/blob/1351254b9d49297d9e72829b6a1109cd18bc054c/Source/WebCore/css/CSSStyleSheet.h#L61

zakariaridouh committed:
Use more smart pointers in CSSStyleSheet
https://bugs.webkit.org/show_bug.cgi?id=278263
<rdar://134095002>

Reviewed by Matthieu Dubet.

Use some more smart pointers in CSSStyleSheet. This is the first of a
series of PRs in which I will adopt more smart pointer usage in respect
to the WebKit Smart Pointer style guide, starting with the `WebCore/css`
folder here.

* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::~CSSStyleSheet):
(WebCore::CSSStyleSheet::RuleMutationScope::RuleMutationScope):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/StyleSheet.h:
(WebCore::StyleSheet::parentStyleSheet const):
(WebCore::StyleSheet::media const):
(WebCore::StyleSheet::ownerRule const):

Canonical link: https://commits.webkit.org/293318@main
